Documentation/Evolution/StringProcessingAlgorithms.md

```diff
@@ -0,0 +1,402 @@
+
+```swift
+extension BidirectionalCollection where SubSequence == Substring {
+    public func ranges<R: RegexProtocol>(of regex: R) -> some Collection<Range<Index>>
+}
+```
+
+### Use custom parsers in regex builders and `RegexProtocol` algorithms
+
+We want to extend string processing to types from outside the standard library, so that one can use custom parsers with the string processing algorithms and incorporate custom parsers in regex builders seamlessly. 
+
+Consider parsing an HTTP header to capture the date field as a `Date` type:
+
+```swift
+let header = """
+HTTP/1.1 301 Redirect
+Date: Wed, 16 Feb 2022 23:53:19 GMT
+Connection: close
+Location: https://www.apple.com/
+"""
+```
+
+You are likely going to match a substring that look like a date string (`16 Feb 2022`), and parse the substring as a `Date` with one of the date parsers in the Foundation framework:
+
+```swift
+let regex = Regex {
+    capture {
+        oneOrMore(.digit)
+        " "
+        oneOrMore(.word)
+        " "
+        oneOrMore(.digit)
+    }
+}
+
+let dateParser = Date.ParseStrategy(format: "\(day: .twoDigits) \(month: .abbreviated) \(year: .padded(4))"
+if let dateMatch = header.firstMatch(of: regex)?.0 {
+    let date = try? Date(dateMatch, strategy: dateParser)
+}
+```
+
+This works, but wouldn't it be much more approachable if you can directly use the date parser within the string match function?
+
+```swift
+let regex = Regex {
+    capture(dateParser)
+}
+
+let date = header.firstMatch(of: regex).map(\.result.1) 
+// A `Date` representing 2022-02-16 00:00:00 +0000
+```
+
+Or consider parsing a bank statement to record all the monetary values in the last column:
+
+```swift
+let statement = """
+CREDIT    04/06/2020    Paypal transfer    $4.99
+CREDIT    04/03/2020    Payroll            $69.73
+DEBIT     04/02/2020    ACH transfer       ($38.25)
+DEBIT     03/24/2020    IRX tax payment    ($52,249.98)
+"""
+```
+
+We have already seen that parsing a date string can be tricky since it could contain localized month name (`"Feb"` as seen from above). Parsing a currency string such as `$3,020.85` with regex is not trivial either -- it can contain grouping separators, a decimal separator, and a currency symbol, all of which can be localized. 
+
+Foundation has various parsers for localized strings like these. Delegating this task to dedicated parsers alleviates the need to handle it yourself. In the second part of the pitch, we introduce the `CustomRegexComponent` protocol that conveniently lets types from outside the standard library participate in regex builders and `RegexProtocol` algorithms.
+
+If `Foundation.FloatingPointFormatStyle<Double>.Currency` conformed to `CustomRegexComponent`, you would be able to retrieve the currency from the bank statement above as a list of `Double` values this way:
```

Comment:
```suggestion
`CustomRegexComponent` allows types, such as `Foundation.FloatingPointFormatStyle<Double>.Currency`, to be used directly within a regex.
```

I understand that you're not proposing the conformances for Foundation here, but I don't think the "if ... you would be able to..." adds clarity.